build!: Remove non-Linux binaries from the image and fix image building on Mac ARM

terrytangyuan opened the pull request:
Currently, this is hard-coded to work with `darwin-amd64` only.

This also fixes the error `#49 52.15 cmd/go: unsupported GOOS/GOARCH pair windows/arm64` as well as the following error when building the UI:

```
 > [argocd-ui 6/6] RUN NODE_ENV='production' NODE_ONLINE_ENV='online' yarn build:
#43 0.399 yarn run v1.22.4
#43 0.420 $ find ./dist -type f -not -name gitkeep -delete && webpack --config ./src/app/webpack.config.js
#43 1.086 Bundling in production...
#43 164.3 Killed
#43 164.7 error Command failed with exit code 137.
#43 164.7 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
------
executor failed running [/bin/sh -c NODE_ENV='production' NODE_ONLINE_ENV='online' yarn build]: exit code: 137
make: *** [image] Error 1
```

The root causes are:
* `GOOS`  and `GOARCH`  were not passed to `make argocd-all` explicitly and `windows + arm64` doesn't work. 
* `NODE_OPTIONS` is required to avoid out-of-memory issue `exit code: 137` when running `yarn build`.

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

terrytangyuan commented:
Failed build for `Deploy` is unrelated and I am fixing that in https://github.com/argoproj/argo-cd/pull/7678 separately as it blocks other PRs as well.

alexmt reviewed (commented):
Thank you for pushing it @terrytangyuan !

Noted an issue with `make image DEV_IMAGE=true` command. I think we can find workaround but there is also a bigger issue. 

The `argocd-darwin-amd64` and `argocd-windows-amd64` are also hardcoded in UI and API server.  https://github.com/argoproj/argo-cd/blob/b37eee1054e42c873699460dd5e2447c2f9fe5a6/ui/src/app/help/components/help.tsx

Binaries are baked into the image so that user can download them using the help page. This is a useful feature for some users however it significantly increases image size for everyone.

I think we can improve it if we stop bundling binaries for all platforms by default and instead make it possible to configure available binaries. Once it is done the main build can just have one image and end users can bundle more binaries using e.g. init containers.

jessesuen commented:
> Binaries are baked into the image so that user can download them using the help page. This is a useful feature for some users however it significantly increases image size for everyone.

Yes, the feature is useful because we even recommend in our docs that the CI pipeline can get the binary at runtime so that they are always using a compatible version and don't have to update the image that the CI uses. 

Some thoughts:
1. we will have either the linux arm64/amd64 baked into the image because, well, argocd uses the same image. So there's no ability to removing it. So i would recommend we link `linux-amd64` `linux-arm64` to the `argocd` binary (whichever one the image was built from). For the opposite architecture, we do not need to bundle it in.
2. Mac, windows, and the opposite linux architecture can be omitted as a default and redirect to GitHub release download link

terrytangyuan commented:
Updated PR with the following changes:
* Removed non-Linux binaries inside the image
* Removed links to non-Linux binaries on help page

These changes will prepare the work in https://github.com/argoproj/argo-cd/issues/7698.

alexmt commented on ui/src/app/help/components/help.tsx:
```diff
@@ -23,14 +23,6 @@ export const Help = () => (
                                 <i className='fab fa-linux ' /> Linux
                             </a>
-                            &nbsp;
-                            <a href={`download/argocd-darwin-amd64`} className='user-info-panel-buttons argo-button argo-button--base'>
```
The `argocd-darwin-amd64` and `argocd-windows-amd64.exe` and mentioned in https://github.com/argoproj/argo-cd/blob/09c5d5f42abe2a8a7a36946cd69c15c625ec98ae/server/server.go#L811

Please remove it as well.

alexmt replied:
Also `server.go` should register `argocd-linux-amd64` or `argocd-linux-arm` based on architecture type. I think we can inject architecture using `LDFLAGS` here: https://github.com/argoproj/argo-cd/blob/09c5d5f42abe2a8a7a36946cd69c15c625ec98ae/Makefile#L131

terrytangyuan replied:
Good call. I used  `runtime.GOARCH` instead of `LDFLAGS` since it's simpler.

alexmt commented on ui/src/app/help/components/help.tsx:
```diff
@@ -23,14 +23,6 @@ export const Help = () => (
                             <a href={`download/argocd-linux-amd64`} className='user-info-panel-buttons argo-button argo-button--base'>
```
Depending on image type download URL should be `argocd-linux-amd64` or `argocd-linux-arm`. You can inject architecture into global variables during build using webpack. See example here:

https://github.com/argoproj/argo-cd/blob/09c5d5f42abe2a8a7a36946cd69c15c625ec98ae/ui/src/app/webpack.config.js#L63

terrytangyuan replied:
Updated but I've never done anything with node so please help double check.

alexmt commented on Dockerfile:
```diff
@@ -115,12 +115,6 @@ COPY . .
 COPY --from=argocd-ui /src/dist/app /go/src/github.com/argoproj/argo-cd/ui/dist/app
 RUN make argocd-all
 
-ARG BUILD_ALL_CLIS=true
```
Sorry, I did not catch it earlier. The Makefile has `release-cli` target to expects these binaries in image: 

https://github.com/argoproj/argo-cd/blob/8bbf8887d25c1c200e88ff3e8ec228d4cd637599/Makefile#L215-L219

Binaries are uploaded to github by by release workflow

https://github.com/argoproj/argo-cd/blob/8bbf8887d25c1c200e88ff3e8ec228d4cd637599/.github/workflows/release.yaml#L191


Can you move  `make BIN_NAME=argocd-darwin-amd64 GOOS=darwin argocd-all && make BIN_NAME=argocd-windows-amd64.exe GOOS=windows argocd-all ` to `release-cli` target please?

terrytangyuan replied:
Good catch. I changed this to an env var and it now defaults to `false` except for `make release-cli` in the release workflow.

alexmt reviewed (approved):
LGTM
